### PR TITLE
Fold better; fix foldl1 bug

### LINKED ISSUE #86: foldl1 looks flat-out wrong
I'm working on a patch that will replace the folds anyway, but the current code,

```haskell
  foldl1 f a | sz == 0   = die "foldl1" "empty array"
             | otherwise = go $ sz-1
   where sz = sizeofArray a
         z = indexArray a 0
         go i | i < 1     = f (go $ i-1) (indexArray a i)
              | otherwise = z
```

doesn't look remotely right. The `i < 1` should be `i > 0` or (probably better) `i /= 0`, or the branches should be swapped, or something.

### PR DESCRIPTION
**Pull request by treeowl**
* Index `Array`s and `SmallArray`s eagerly.
* Fix `foldl1` for `Array`, which was completely wrong.
* Make the structure of the `Foldable` instances for `Array`
  and `SmallArray` the same, taking what I thought was the best
  from each.

Fixes #86

### CODE REVIEW

**Comment by cartazio** [comment]
Look pretty solid.

Could you explain what you mean by structure of the foldable instances ?
The code looks clean and good, I’m just not sure if you mean formatting of
instances or semantics :)

**Comment by treeowl** [comment]
Oh, I just meant the way the code is written. `fix` vs explicit recursion, the way the first/last element is handled in `foldr1`/`foldl`. I also removed the explicit `toList` implementations because they really just did the same as the default ones in the same way, which seemed a bit silly.

File: Data/Primitive/SmallArray.hs [hunk 1]
```diff
@@ -424,53 +434,96 @@ instance Ord a => Ord (SmallArray a) where
...
+  minimum ary | sz == 0   = die "minimum" "Empty SmallArray"
+              | (# frst #) <- indexSmallArray## ary 0
+              = go 1 frst
+   where sz = sizeofSmallArray ary
+         go i !e
+           | i == sz = e
+           | (# x #) <- indexSmallArray## ary i
+           = go (i+1) (min e x)
+  {-# INLINE minimum #-}
+  sum = foldl' (+) 0
```
Comment (RyanGlScott) [inline Data/Primitive/SmallArray.hs]: Out of curiosity, does this generate better code than the default implementation for `sum` (which is something like `getSum . foldMap Sum`)?

**Comment by treeowl** [comment]
I think we can merge this now. I don't want to have to keep rebasing the
traversal stuff over it.

**Comment by treeowl** [comment]
Thanks.

**Comment by andrewthad** [comment]
For posterity's sake, I'll point out that `foldl1` is incorrect in this PR but has been fixed in https://github.com/haskell/primitive/pull/99.

**Comment by treeowl** [comment]
I'm surprised. I did test it manually with the list monoid. Maybe I did
something stupid afterwards?

**Comment by andrewthad** [comment]
Maybe you accidentally picked a list that was the reversal of itself. Or maybe you accidentally revert part of your changes before committing. Either way, not a big deal now that we have decent tests.

**Pull request merged**